File: Day 8/Minimum Window Substring.cpp
using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        if (t.size() > s.size()) return "";

        int need[128] = {0}, window[128] = {0};
        int needKinds = 0;


        for (char c : t) {
            if (need[(unsigned char)c] == 0) needKinds++;
            need[(unsigned char)c]++;
        }

        int formed = 0;
        int bestLen = INT_MAX;
        int bestL = 0;
        int l = 0;

        for (int r = 0; r < (int)s.size(); ++r) {
            unsigned char c = s[r];
            window[c]++;

            if (need[c] > 0 && window[c] == need[c]) {
                formed++;
            }


            while (formed == needKinds) {
                if (r - l + 1 < bestLen) {
                    bestLen = r - l + 1;
                    bestL = l;
                }
                unsigned char lc = s[l];
                window[lc]--;
                if (need[lc] > 0 && window[lc] < need[lc]) {
                    formed--;
                }
                l++;
            }
        }

        return (bestLen == INT_MAX) ? "" : s.substr(bestL, bestLen);
    }
};

File: Day 8/Minimum Window Substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""
        need = Counter(t)
        window = {}

        have, need_count = 0, len(need)
        res,res_len = [-1,-1], float("inf")
        left = 0
        for right,char in enumerate(s):
            window[char] = window.get(char,0) +1

            if char in need and window[char] == need[char]:
                have +=1
            while have == need_count:
                if (right - left -1) < res_len:
                    res = [left,right]
                    res_len = right - left -1

                window[s[left]] -= 1
                if s[left] in need and window[s[left]] < need[s[left]]:
                    have -=1
                left +=1
        l,r = res
        return s[l:r+1] if res_len != float("inf") else ""
